feat(iota-json-rpc-api, ts-sdk): Allow to pass `IotaObjectDataOptions` to the `get_dynamic_field_object` function

samuel-rufi opened the pull request:
# Description of change

Allows to provide `IotaObjectDataOptions` to the `get_dynamic_field_object` endpoint.

## Links to any relevant issues

Fixes https://github.com/iotaledger/iota/issues/5807
Fixes https://github.com/iotaledger/iota/issues/6116

## Type of change

- Breaking change

## How the change has been tested

Tested in the dedicated PRs.

Commit:
feat(iota-json-rpc-api)!: Add `IotaObjectDataOptions` to the `get_dynamic_field_object` fn signature (#6095)

* feat: Add `IotaObjectDataOptions` to `get_dynamic_field_object` function signature

* refactor: Add missing arg

* refactor: Fix typo

* refactor: Use default impl for `IotaObjectDataOptions` if arg is `None`

* refactor: Use full content options for compatibility and to avoid breaking changes

msarcev committed:
chore(ts-sdk): Sync TS SDK with changes on get_dynamic_field_object (#6211)

# Description of change

* Regenerated ts-sdk types from openrpc.json
* Since its an optional field in an existing params definition for an
existing method, no additional changes to client needed

## Links to any relevant issues

Closes https://github.com/iotaledger/iota/issues/6116

samuel-rufi commented:
Needs to be rebased as of the latest develop changes in [validator_commands.rs](https://github.com/iotaledger/iota/blob/develop/crates/iota/src/validator_commands.rs) with the clippy fix on top.

Thoralf-M commented on crates/iota-indexer/src/apis/indexer_api.rs:
```diff
@@ -248,7 +249,8 @@ impl IndexerApiServer for IndexerApi {
         .expect("deriving dynamic field id can't fail");
 
-        let options = iota_json_rpc_types::IotaObjectDataOptions::full_content();
+        let options = options.unwrap_or_else(IotaObjectDataOptions::full_content);
```
For get_object() it just uses `let options = options.unwrap_or_default();` which has set all the options to false, just wondering if it would make more sense to do the same here? Otherwise default will require much more data

samuel-rufi replied:
I see what you mean but then we have a breaking change. Each dep that makes use of the current `get_dynamic_field_object` function expects the full content as that was the default behavior for this function before. Shouldn't we avoid changing that? What do you think? @Thoralf-M @thibault-martinez

Thoralf-M replied:
Ah ok wasn't aware that it was like that before also. So it returns everything except bcs and display, I thought full_content also includes these
Then it should be fine as it is :+1:

Commit:
fix(iota): Fix `IotaTransactionBlockResponseOptions` param for `get_dynamic_field_object` in `validator_commands.rs`. (#6325)

# Description of change

With latest changes on develop, the
`IotaTransactionBlockResponseOptions` param for
`get_dynamic_field_object` in `validator_commands.rs` needs to be fixed
in https://github.com/iotaledger/iota/pull/6316.

## Relevant Links

https://github.com/iotaledger/iota/pull/6316#issuecomment-2787410827

## Type of change

- Bug fix (a non-breaking change which fixes an issue)

## How the change has been tested

`cargo build`
`cargo clippy`

## Change checklist

- [X] I have followed the contribution guidelines for this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my
feature works
- [X] I have checked that new and existing unit tests pass locally with
my changes

samuel-rufi commented:
It appears we should better have a dedicated method instead of modifying the signature of the current one to avoid any potential breaking changes.

samuel-rufi commented:
To avoid any potential breaking changes it appears to be safer to introduce a separate `_v2` method. Even though the function behavior stays the same and TS dependencies can handle the optional value, external `IotaClient` usages need to be updated. This `_v2` method will be introduced as part of a dedicated PR for easier coordination.
